Bug : Use random IV in EncryptDecryptUsingTpm

shjala opened the pull request:
# Description

https://github.com/lf-edge/eve/pull/5494/changes/265c6889d1ffa298f6b13cbf7cd5fc4012222d89 Changes EncryptDecryptUsingTpm function to support versioned encryption
with proper random IV generation. Previously, the function used a static
zero IV which is insecure. This commit:

- Adds VaultKeyEncVersion enum (Unespecified, EncryptionAEAD) to track
  encryption method versions
- Implements AES-GCM (AEAD) encryption/decryption with random nonce
- Maintains backward compatibility with existing encrypted data using
  Unespecified version

The new AEAD mode generates a random nonce per encryption operation and
prepends it (with its size) to the ciphertext, providing authenticated
encryption with proper IV randomness.

https://github.com/lf-edge/eve/pull/5494/changes/f32bad68cd5f55dc4c189f875206049c57d36781 Refactor TPM tests and add encryption unit tests
- Move GetCertHash to evetpm package for better reusability.
- Migrate tpmmgr internal tests (testTpmEcdhSupport, testEcdhAES) to evetpm unit tests.
- Add unit tests in encryptdecrypt_test.go covering AES GCM and TPM encryption.
- Update TPM test setup scripts (tests/tpm/*.sh) to provision required keys and certificates.
- Allow overriding of TPM file paths in locationconsts.go for testing purposes.

## PR dependencies
None.

## How to test and validate this PR

### TPM Encryption Fix Validation

Ensure that device encryption keys (Vaults) continue to work correctly for new installations AND after upgrading existing devices.

#### 1. Fresh Installation Verification
Ensure the new encryption method works on a new device.

*   **Steps:**
    1.  **Onboard** a device with this new EVE image.
    2.  Wait for the device to become **Online** in the Controller.
    3.  **Deploy** a generic Edge Application with encrypted volume
    4.  **Verify:**
        *   Does the vault unlocks?
        *   Does the App instance reach **RUNNING** state? (Success)
        *   Does the App instance get stuck in **BOOTING** or **downloading** with errors? (Failure)
    6.  **Log Check**:
        *   Check `tpmmgr` logs on the device or via Controller:
            *   *Search for:* "TPM Sanity Check"
            *   *Expected:* No errors. If you see "failed to encrypt" or "failed to decrypt", the test failed.

#### 2. Upgrade & Backward Compatibility
Ensure that updating an OLD device to this NEW version does not lock the user out of their existing encrypted data.

*   **Steps:**
    1.  **Flash/Install** an **OLD** stable version of EVE (e.g., previous release) on a TPM-capable device.
    2.  **Onboard** and **Deploy** an Edge App with encrypted volume.
    3.  Verify the App is **RUNNING**.
    4.  **Trigger an Update** to the **NEW** EVE image (with this PR include).
    5.  Wait for the device to update and reboot.
    7.  **Verify:**
        *   Does the vault unlocks and *existing* App instance start up and reach **RUNNING** state again?
        *   The old vault's keys are stored on the Controller using the OLD encryption format. The new EVE image must correctly detect and decrypt these "Legacy" keys.

## Changelog notes

Fix AES-CFB encryption in EVE by by moving into AES-GCM (AEAD) with random nonce. ⚠️ This change breaks downgrade compatibility (downgrades are not officially supported). Once an upgrade is successful, the vault key format stored in the controller is changed. Any EVE version that does not include this change will be unable to decrypt the vault key sent by the controller.

## PR Backports

For all current LTS branches, please state explicitly if this PR should be
backported or not. This section is used by our scripts to track the backports,
so, please, do not omit it.

Here is the list of current LTS branches (it should be always up to date):

- 14.5-stable
- 13.4-stable

For example, if this PR fixes a bug in a feature that was introduced in 14.5,
you can write:

```text
- 14.5-stable: To be backported.
- 13.4-stable: No, as the feature is not available there.
```

Also, to the PRs that should be backported into any stable branch, please
add a label `stable`.

## Checklist

- [ ] I've provided a proper description
- [ ] I've added the proper documentation
- [ ] I've tested my PR on amd64 device
- [ ] I've tested my PR on arm64 device
- [ ] I've written the test verification instructions
- [ ] I've set the proper labels to this PR

For backport PRs (remove it if it's not a backport):

- [ ] I've added a reference link to the original PR
- [ ] PR's title follows the template

And the last but not least:

- [ ] I've checked the boxes above, or I've provided a good reason why I didn't
  check them.

Please, check the boxes above after submitting the PR in interactive mode.

eriknordmark reviewed (commented):
This assumes that the controllers do not care about the length of the encrypted blob. Need to check whether that is the case. Does it place any other requirements on the controllers?

Also, presumably the decryption needs to handle the case when there is no prepended IV and use the old zeros approach so that one can transition.
(When it does that, does it always send the new encrypted blob to the controller meaning the IV would be added on the next boot of a device?)

Last but nost least, the API proto files which describe this as an encrypted key should have their comments expanded; key thing is to not assume anything about the length of the blob since the blob may or may not include an IV (for new vs old EVE versions).

justincormack commented:
You are making this very hard to change in the future, you should properly version the new interface. CFB encryption is no longer recommended (as it is not authenticated), and CFB modes are marked as deprectaed in Go standard library, so you are going to have to change this again, so you might as well make this easier.

shjala committed:
Fix: Use random IV in TPM encryption with AEAD mode

Changes EncryptDecryptUsingTpm function to support versioned encryption
with proper random IV generation. Previously, the function used a static
zero IV which is insecure. This commit:

- Adds VaultKeyEncVersion enum (Unespecified, EncryptionAEAD) to track
  encryption method versions
- Implements AES-GCM (AEAD) encryption/decryption with random nonce
- Maintains backward compatibility with existing encrypted data using
  Unespecified version

The new AEAD mode generates a random nonce per encryption operation and
prepends it (with its size) to the ciphertext, providing authenticated
encryption with proper IV randomness.

Signed-off-by: Shahriyar Jalayeri <shahriyar@posteo.de>

shjala committed:
Refactor TPM tests and add encryption unit tests

- Move GetCertHash to evetpm package for better reusability.
- Migrate tpmmgr internal tests (testTpmEcdhSupport, testEcdhAES) to evetpm unit tests.
- Add unit tests in encryptdecrypt_test.go covering AES GCM and TPM encryption.
- Update TPM test setup scripts (tests/tpm/*.sh) to provision required keys and certificates.
- Allow overriding of TPM file paths in locationconsts.go for testing purposes.

Signed-off-by: Shahriyar Jalayeri <shahriyar@posteo.de>

eriknordmark reviewed (approved):
LGTM but some revive and codespell issues to fix.

shjala committed:
Increase timeout in evaluation cycle tests

Increase timeout durations for stability check and inventory collection
in evaluation cycle tests.

Signed-off-by: Shahriyar Jalayeri <shahriyar@posteo.de>

eriknordmark commented:
Running against the commercial controller I see failures. The system updates fine to run the image in this PR, and as part of that it saves the encrypted key in the controller.
But the tests which try the EVE image update (by going to e.g., current master build) fail with
ERROR: device: baseos_updating attest: InternalEscrowWait error [ATTEST] No escrow data vault: ERROR error Vault key unavailable 

Maybe this is a constraint but it means there will be issues if update to a newer version is done, and it saves the key in the new format with an IV, and then the system crashes during its first 10 minutes and we fallback to the previous version.
That previous version does not know to extract the IV from the new format so it fails??

Do we need to introduce this using two steps?
1. Make EVE version N be able to parse the new format and use the IV
2. Wait for weeks or months
3. Make EVE version N+k save the key with a non-empty IV

shjala commented:
> Do we need to introduce this using two steps?

Let me think about it a bit, but maybe we have to.

eriknordmark commented:
> @eriknordmark What do you think about this strategy :
> 
> **_Boot 1_** : EVE is potentially unstable (PartitionState = inprogress), encrypt using AES-ZeroIV (LEGACY) regardless of what EVE version is running, Controller receives legacy data. If rollback happens, Controller has LEGACY data that every EVE version can understand.
> 
> **_Boot 2_** : EVE is stable (PartitionState = active), encrypt using AEAD, Controller receives new data. Rollback is not going to happen and we are committed to the latest version.
> 
> This relies on no disk write and we don't have to worry about disk fails. We can indicate in the release notes that upgrading to this version makes rollback not only officially but also technically impossible without losing vault data.

That sounds workable and simple enough.

Please make sure you put the release note text in the PR description, and let me know when you have updated the code so we can run this through the tests.

We also need to check how we test things - in eden and in Zededa test setups to make sure we only test upgrades (from LTS releases to current PR or current master/release) and not try to move back and forth across releases.
Can you check how Eden does the EVE update tests?

eriknordmark commented on pkg/pillar/cmd/vaultmgr/vaultmgr.go:
```diff
@@ -538,8 +539,15 @@ func publishVaultKey(ctx *vaultMgrContext, vaultName string) error {
-		// if this fails, tpm manager signals the controller something is wrong with TPM
-		encryptedKey, err := etpm.EncryptDecryptUsingTpm(keyBytes, true)
+		// If EVE is potentially unstable (PartitionState = inprogress), encrypt using
+		// AES-ZeroIV (LEGACY) regardless of what EVE version is running.
```
Makes sense to add that this is needed so that if there is a fallback to the old version it will be able to decrypt using the old (unspecified = zero IV) method.

But, this needs some more checks because we don't want this when updating from a future 17.0.0 to 18.0.0. Can we track that we already encypted using AEAD and never go back (and ideally, not allow a downgrade attack by editing some file in /persist).
Would a file like /config/persist_first_aead_version containing "16.9.0" written the first time you use AEAD accomplish this?

If so, we could (separately) add a checkin baseosmgr that if the semver of the new EVE version is not 0.0.0 and is less than the string in persist_first_aead_version), then baseosmgr should generate an error insteaf of trying the update..

rene reviewed (changes requested):
@shjala , @eriknordmark I'm marking Request changes since this PR has now several conflicts to be solved. Also, with the fallback mechanism we won't have any issues across EVE version, is my understanding right?
